Remove unused argument from GetRegionDisplayName

stephentoub opened the pull request:
cc: @tarekgh

stephentoub committed:
Remove unused argument from GetRegionDisplayName

jkotas commented on src/System.Private.CoreLib/shared/System/Globalization/CultureData.cs:
```diff
@@ -1054,17 +1054,14 @@ internal string LocalizedCountryName
-                    {
-                        _sLocalizedCountry = NativeCountryName;
-                    }
+                    _sLocalizedCountry ??= NativeCountryName;
```
Is this method supposed to be thread safe - do we have a race condition here?

stephentoub replied:
More so than all of the other lazily-initialized members on this type?

Most of these cases seem to be based on the idea that the value multiple threads would produce is idempotent, and thus the race wouldn't matter.  This one is a bit odd, in that GetRegionDisplayName factors in the current thread's culture.

jkotas replied:
> Most of these cases seem to be based on the idea that the value multiple threads would produce is idempotent, and thus the race wouldn't matter. 

Right. The field has to be set just once to the final value for this to work. This one sets it twice - first to `null` and second time to `NativeCountryName`. The other racing thread can return `null` from this method as a result.

> GetRegionDisplayName factors in the current thread's culture, and thus isn't necessarily.

I have not even noticed this problem...

stephentoub replied:
> This one sets it twice 

Yup.  There are actually several cases of this in this file.  I'll fix them all.

stephentoub replied:
I pushed a commit to address the race conditions, and opened https://github.com/dotnet/coreclr/issues/27808 for the potential functional issue.

stephentoub committed:
Address race conditions in CultureData

While most of the race conditions in this file are benign, in these cases, a thread could end up returning an intermediate value due to multiple potential writes to the same field.

tarekgh reviewed (commented):
LGTM
